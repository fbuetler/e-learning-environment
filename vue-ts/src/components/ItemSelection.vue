<template>
  <div class="flex-item flex-wrap flex-row">
    <div
      class="flex-item flex-wrap flex-center card clickable"
      v-for="item in items"
      :key="item.id"
      :class="[
        {
          selected: item.type === selected,
        },
        item.class,
      ]"
      @click="$emit('selected', item.type)"
      draggable
      @dragstart="$emit('selected', item.type)"
      @dragover.prevent
      @dragend.prevent
      @drop.prevent
    >
      <img :src="require('@/assets/' + item.img)" />
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop } from "vue-property-decorator";

export type item = {
  id: number;
  type: number;
  value: number;
  img: string;
  class: string;
};

@Component<ItemSelection>({})
export default class ItemSelection extends Vue {
  @Prop({ required: true })
  selected: number;
  @Prop({ required: true })
  items: item[];
}
</script>
