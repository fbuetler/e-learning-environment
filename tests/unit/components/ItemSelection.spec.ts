import { shallowMount } from "@vue/test-utils";
import ItemSelection from "@/components/shared/ItemSelection.vue";

describe("ItemSelection.vue", () => {
  let wrapper;
  const selected = 1;
  const items: Array<{
    id: number;
    type: number;
    value: number;
    img: string;
    class: string;
  }> = [
    {
      id: 1,
      type: 1,
      value: 42,
      img: "beavers/beavers.png",
      class: "",
    },
    {
      id: 2,
      type: 3,
      value: 27,
      img: "text",
      class: "duda",
    },
  ];
  beforeEach(() => {
    wrapper = shallowMount(ItemSelection, {
      propsData: {
        selected: selected,
        items: items,
      },
    });
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("image items are rendered correctly", async () => {
    const imgItems = wrapper
      .find(`#item-selection-${items[0].id}`)
      .element.getElementsByTagName("img");
    expect(imgItems).toBeTruthy();
    expect(imgItems.length).toBe(1);
  });

  it("text items are rendered correctly", () => {
    const divItems = wrapper
      .find(`#item-selection-${items[1].id}`)
      .element.getElementsByTagName("div");
    expect(divItems).toBeTruthy();
    expect(divItems.length).toBe(1);
  });

  for (let i = 0; i < items.length; i++) {
    it(`event triggered if clicked on item ${i}`, async () => {
      await wrapper.find(`#item-selection-${items[i].id}`).trigger("click");
      expect(wrapper.emitted("selected")).toBeTruthy();
      expect(wrapper.emitted("selected").length).toBe(1);
      expect(wrapper.emitted("selected")[0]).toEqual([items[i].type]);
    });
  }

  for (let i = 0; i < items.length; i++) {
    it(`event triggered if dragstart item ${i}`, async () => {
      await wrapper.find(`#item-selection-${items[i].id}`).trigger("dragstart");
      expect(wrapper.emitted("selected")).toBeTruthy();
      expect(wrapper.emitted("selected").length).toBe(1);
      expect(wrapper.emitted("selected")[0]).toEqual([items[i].type]);
    });
  }

  it("only selected item is highlighted", async () => {
    const selectedItem = wrapper.find(`#item-selection-${selected}`);
    expect(selectedItem.element.className).toContain("selected");
    const selectedItems = wrapper.findAll(".selected");
    expect(selectedItems.length).toBe(1);
  });
});
