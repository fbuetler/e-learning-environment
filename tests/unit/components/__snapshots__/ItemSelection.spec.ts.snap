// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemSelection.vue renders correctly 1`] = `
<div
  class="flex-item flex-wrap flex-center-horizontally flex-row"
  id="item-selection"
>
  <div
    class="flex-item flex-wrap flex-center card clickable responsive selected"
    draggable="true"
    id="item-selection-1"
  >
    <img
      src=""
    />
  </div>
  <div
    class="flex-item flex-wrap flex-center card clickable responsive duda"
    draggable="true"
    id="item-selection-2"
  >
    <div>
      text
    </div>
  </div>
</div>
`;
